corehq/apps/app_manager/app_schemas/case_properties.py

```diff
@@ -18,117 +19,118 @@
+            forms.append(form)
+    return forms
+
+
 class ParentCasePropertyBuilder(object):
 
     def __init__(self, app, defaults=(), per_type_defaults=None):
         self.app = app
         self.defaults = defaults
         self.per_type_defaults = per_type_defaults or {}
 
-    @property
+    def _get_relevant_apps(self, include_shared_properties):
+        apps = [self.app]
+        if self.app.case_sharing and include_shared_properties:
+            for app in self._get_other_case_sharing_apps_in_domain():
+                apps.append(app)
+        return apps
+
     @memoized
-    def forms_info(self):
-        # unfortunate, but biggest speed issue is accessing couchdbkit properties
-        # so compute them once
-
-        forms_info = []
-        if self.app.doc_type == 'RemoteApp':
-            return forms_info
-        for module in self.app.get_modules():
-            for form in module.get_forms():
-                forms_info.append((module.case_type, form))
-        return forms_info
-
-    @property
+    def _get_relevant_forms(self, include_shared_properties):
+        forms = []
+        for app in self._get_relevant_apps(include_shared_properties):
+            forms.extend(_get_forms(app))
+        return forms
+
     @memoized
-    def case_sharing_app_forms_info(self):
-        forms_info = []
-        for app in self.get_other_case_sharing_apps_in_domain():
-            for module in app.get_modules():
-                for form in module.get_forms():
-                    forms_info.append((module.case_type, form))
-        return forms_info
+    def _get_all_case_updates(self, include_shared_properties):
+        all_case_updates = defaultdict(set)
+        for form in self._get_relevant_forms(include_shared_properties):
+            for case_type, case_properties in form.get_all_case_updates().items():
+                all_case_updates[case_type].update(case_properties)
+        return all_case_updates
 
     @memoized
-    def get_parent_types_and_contributed_properties(self, case_type, include_shared_properties=True):
-        parent_types = set()
-        case_properties = set()
-        forms_info = self.forms_info
-        if self.app.case_sharing and include_shared_properties:
-            forms_info += self.case_sharing_app_forms_info
+    def _get_data_dictionary_properties_by_case_type(self):
+        return {
+            case_type: {prop.name for prop in props} for case_type, props in groupby(
+                CaseProperty.objects
+                .filter(case_type__domain=self.app.domain, deprecated=False)
+                .order_by('case_type__name'),
+                key=attrgetter('case_type.name')
+            ).items()
+        }
+
+    def get_case_relationships_for_case_type(self, case_type, include_shared_properties=True):
+        return self.get_case_relationships(include_shared_properties)[case_type]
+
+    @memoized
+    def get_case_relationships(self, include_shared_properties=True):
```

Comment:
Would be nice for all (at least non-private) functions like this to have a docstring that describes the data structure of the returned object. For example, this one might be
```
Returns a `defaultdict(set)` where each key is a case type
(string) and each value is a set of ...
```
I'm not even sure how to finish that, but guessing it's a set of relationship names (strings) like `"parent"`.